### ISSUE TITLE
Loss of precision (especially when trading ratios are high)

### ISSUE THREAD

**Issue by adrianhopebailie** [interledger/rfcs#354]
If a multi-hop payment is routed over a ledger that supports less relative precision than that used by the initial transfer then it seems likely that some precision is lost for all transfers further along the route.

Example: Alice sending USD to Bob (via Chloe and Chris who use BTC)
1. Alice sends $ 0.1234 to Chloe
2. Chloe converts to BTC and wants to send 0.00000822666666 (at a rate of $15K/BTC) to Chris but Bitcoin only supports precision to 8 decimals so Chloe sends 0.00000822 (0.00000823 would mean Chloe makes a loss)
3. Chris converts from 0.00000822 BTC to $0.1233 USD and sends this to Bob.

In a situation where the connectors intended to charge no fee, in effect Chloe has taken a fee ($0.0001) because the ledger she shares with Chris forced her to round down and claim the remainder.

This is despite that ledger supporting very high precision because the rate of exchange between the two ledgers is so high.

If the accepted precision of a currency is `x` then a currency that trades at `1:10^y` against that currency MUST support a precision of `>= (x + y)` for these to be tradable at low values without losing precision.

So, if we expect most USD ledgers to support a precision of 2 and BTC has a precision of 8 then we better hope the value of BTC doesn't go above $1000000 😄 

But, if we want to have USD ledgers that have higher precision (let's say 4 as in the example above) then BTC is already a problematic bridge asset for a USD micropayment because it's trading at > $10000.

Is it sufficient for us to say that there is an incentive for connectors to find ledgers that support high levels of precision in order to offer more competitive rates?

Does the protocol need a way to probe for the max precision along a route (similar to max bandwidth as discussed by @emschwartz)?

**Comment by michielbdejong** [comment]
This is not really a problem because it only affects the amount escrowed in on-ledger escrow; Chloe and Chris can use out-of-band communication to pass the more precise amount. Same for payment channels, the updated claim would be in the ledger precision, but the transfer can have a higher precision.

Note that the example of BTC is not realistic because the 0.0001 or 0.01 USD rounding error on the escrow amount would be dwarfed by a BTC on-ledger transaction fee of over 10 USD

**Comment by adrianhopebailie** [comment]
You are right that this issue is not caused by limitations of the ledger but it could still manifest if Chris and Chloe have setup a channel without the necessary precision.

They also take on risk themselves by performing transfers using a channel that has greater precision than the underlying ledger. i.e. What is the rounding rule when it's time to close the channel? How do you submit the claim to the ledger with more precision than it supports?

> Note that the example of BTC is not realistic because the 0.0001 or 0.01 USD rounding error on the escrow amount would be dwarfed by a BTC on-ledger transaction fee of over 10 USD

Except that off-ledger transfers can be free

**Comment by millerabel** [comment]
The MaxPrecision probe is an interesting idea. If this were incorporated into route quoting, it would allow for channel selection based on precision, as well as direct fees, bandwidth, and other channel facets.

**Comment by dora-gt** [comment]
- Is it sufficient for us to say that there is an incentive for connectors to find ledgers that support high levels of precision in order to offer more competitive rates?
  - My answer is Yes because connectors can not exist without ILP economy. ILP economy can continue only if people use ILP, gaining benefit (both cost and experience) from it. So connectors SHOULD choose the best ledger (costs less, and enables micro-payments) for people especially in the case that people do micro-payments. Am I idealist? 😅
- Does the protocol need a way to probe for the max precision along a route (similar to max bandwidth as discussed by @emschwartz)?
  - This is just my opinion. I think No because the total cost really, only matters.
  - While if people want to know the reason "why fee is high", we may need the probe function.

**Comment by sharafian** [comment]
The connector always rounds in their favor, so losing money wouldn’t be a concern in this scenario. It would just make the path look a little more expensive from the user’s perspective. And because ILP doesn’t go on-ledger anymore, two connectors can always just agree to use a high precision so that their customers don’t get charged for rounding.

**Comment by adrianhopebailie** [comment]
> two connectors can always just agree to use a high precision so that their customers don’t get charged for rounding

Correct, but this may only materialize as an issue when the exchange rate on a route changes to exceed some threshold.

i.e. As you update your routing table you may discover a route where this is a problem. I assume you can't dynamically change the precision on one of the accounts on that route? So what should you do?

**Comment by sharafian** [comment]
The only exchange rates you deal with are local ones. You wouldn't discover a problematic route, just a problematic trading pair. Because you're the one who configures these trading pairs, it's just part of your responsibility as a connector to keep the scales about right. Because the scales go by factors of 10 this is an easy target to hit.

If you're exchanging BTC against USD you probably want them within 5 or 6 orders of magnitude of each other (so that you can trade 1e-9 USD for 1e-15 BTC). This wouldn't encounter any problems unless BTC went to 100k, in which case you'd have an effective scale of 8 for USD rather than 9 on that trading pair. You and your peer can agree to update your configuration, after which the problem will be fixed.

**Comment by adrianhopebailie** [comment]
> The only exchange rates you deal with are local ones. You wouldn't discover a problematic route, just a problematic trading pair.

Trading pair just tells you the rate. A route is the rate for two specific accounts (which each have a scale).

However I'd expect most connectors to not "set exchange rates" but rather base their rates of some widely accepted rate plus their fee. As such, these are constantly changing and it should be noted "somewhere" that as they change the scales in use may become problematic and this is something a good connector should monitor.
